src/games/rcll/custom-orders.clp

```diff
@@ -0,0 +1,272 @@
+  (pb-set-field ?copy "quantity_delivered_magenta" (pb-field-value ?orig "quantity_delivered_magenta"))
+  (pb-set-field ?copy "delivery_period_begin" (pb-field-value ?orig "delivery_period_begin"))
+  (pb-set-field ?copy "delivery_period_end" (pb-field-value ?orig "delivery_period_end"))
+  (pb-set-field ?copy "delivery_gate" (pb-field-value ?orig "delivery_gate"))
+  (pb-set-field ?copy "competitive" FALSE)
+  
+  (return ?copy)
+)
+
+(deffunction get-orderinfo-extid
+  (?orderinfo)
+  
+  (bind ?eid 0)
+  
+  ;(bind ?orders (pb-field-list ?orderinfo "orders"))
+  ;(bind ?o (first$ ?orders)
+  ;(bind ?eid (pb-field-value ?o "id"))
+
+  (foreach ?o (pb-field-list ?orderinfo "orders")
+    (bind ?eid (pb-field-value ?o "id"))
+    (break)
+  )
+    
+  (return ?eid)
+)
+
+;--
+;-- RULES
+;--
+
+(defrule print-custom-order-success
+  (custom-order)
+=>
+  (printout warn "INFO: Enabling custom-order was successful" crlf)
+)
+
+(defrule remove-initial-orders
+  (custom-order)
+  ?of <- (order (id ?id&:(< ?id 10))
+                (start-range ?ss&:(< ?ss 1320) ?se&:(<= ?se 1320))
+         ) 
+=>
+  (printout warn "INFO: Remove initial order id " ?id crlf)
+
+  ;-- remove original game order
+  (retract ?of)  
+)
+
+(defrule net-recv-custom-OrderInfo
+  (custom-order)
+  (gamestate (state RUNNING) (phase PRODUCTION) (game-time ?now))
+  ?pf <- (protobuf-msg (type "llsf_msgs.OrderInfo")
+            (ptr ?ptr)
+            (rcvd-via STREAM)
+            (rcvd-from ?from-host ?from-port)
+            (client-id ?cid)
+         )
+  ;-- reject already processing orders
+  (not (orderer-client-response-info 
+          (extern-id ?eid&:(eq ?eid (get-orderinfo-extid ?ptr)))
+        )
+  )
+  
+=>
+
+  (printout warn "INFO: Received order information from client " ?cid crlf)
+  
+  ;-- insert orders from message
+  (bind ?recv-resp (pb-create "llsf_msgs.Order"))
+  (bind ?eid -1)
+  (bind ?cnt 0)
+  (foreach ?o (pb-field-list ?ptr "orders")
+    (if (< ?eid 0) then
+      (bind ?recv-resp (copy-order ?o))
+      (bind ?eid (pb-field-value ?o "id"))
+    )
+
+    (bind ?id 
+      (assert-custom-order ?o ?now)
+    )
+    
+    ;-- increase product count
+    (bind ?cnt (+ ?cnt 1))
+    
+    ;-- assert orderer-client-response-product-info
+    (bind ?quantity (pb-field-value ?o "quantity_requested"))
+    (assert (orderer-client-response-product-info
+              (extern-id ?eid)
+              (product-id ?id)
+              (quantity ?quantity)
+              (last-delivered 0.0)
+            )
+    )
+    
+    (printout warn "INFO: Append " ?quantity "x of product " ?id " to external order " ?eid crlf)
+  )
+  
+  ;-- remember orders and orderer (client)
+  (assert (orderer-client-response-info
+            (client ?cid) 
+            (extern-id ?eid) 
+            (product-count ?cnt)
+          )
+  )
+    
+  (printout warn "INFO: Inserted external order " ?eid " with " ?cnt " products" crlf)
+  (pb-send ?cid ?recv-resp)
```

Comment:
Why do need to send the OrderInfo immediately? Isn't it enough to just assert the order and then rely on the regular OrderInfo publishing? This would simplify the code

Reply:
Will try that out

Reply:
it's not an OrderInfo message, but a Order message, this is send so that the ROS node knows, that the order was transmitted. I think was implemented to have "simple" way of knowing whether an order has been send to the refbox. Since OrderInfo publishes all current orders and you would have to refind your own order? maybe I'm mistaken

Reply:
Yes, but the `OrderInfo` is just a list of `Order`s. Is there a reason why this `Order` needs to be sent separately?

It's also odd that you receive an OrderInfo and then send back a single Order. Shouldn't it be the other way around? The client sends a single Order to be added. The refbox responds with the list of all orders.

Reply:
Talked to @theclakuh, the OrderInfo is received from the webshop backend (the transformation from the webshop order into an order that can be processed by the refbox), the order that is send back is a confirmation that the order was handled properly by the refbox.